### PR TITLE
Allow all private declarations to be emitted in declaration output

### LINKED ISSUE #23127: Relax visbility rules for declaration emit
Visibility rules should be relaxed to only assert that a non-exported declaration is never exported.

```ts
class TreeNode {}
class Tree {
        // error TS4031: Public property 'Node' of exported class has or is using private name 'TreeNode'.
	Node: typeof TreeNode;
}
namespace Tree {
}
export = Tree;
```

another example:

```ts
interface I {
    a: number;
}

class C {
    // error TS4063: Parameter 'a' of constructor from exported class has or is using private name 'I'.
    constructor (a: I) { }
}

export default C;
```

### PR DESCRIPTION
**Pull request by weswigham**
Followup to #22798 the allows any top-level declaration to be made visible if it is required to properly emit another emitted declaration.

Fixes #23127

Makes #15058 redundant.

As a side effect, because we now consider these declarations "visible", we now sometimes use their names in the `type` baselines.

### CODE REVIEW

**Comment by bluelovers** [comment]
looks like this is something i need it lol

**Comment by weswigham** [comment]
@sheetalkamat Would you like to give this a review? The GH UI won't let me add people T.T

File: tests/baselines/reference/declarationEmitLocalClassDeclarationMixin.js [hunk 1]
```diff
@@ -107,3 +107,4 @@ declare const FilteredThing_base: {
 export declare class FilteredThing extends FilteredThing_base {
     match(path: string): boolean;
 }
+export {};
```
Comment (sheetalkamat) [inline tests/baselines/reference/declarationEmitLocalClassDeclarationMixin.js]: where is this coming from ?

Comment (weswigham) [reply tests/baselines/reference/declarationEmitLocalClassDeclarationMixin.js]:
```
declare const FilteredThing_base: {
     new (...args: any[]): {	     new (...args: any[]): {
         match(path: string): boolean;	         match(path: string): boolean;
         thing: number;	         thing: number;
     };	     };
 } & typeof Unmixed;	 } & typeof Unmixed;
```
doesn't exist at runtime (we just emit it so we have a thing for the `extends` clause) - this scopes the file so it's actually private.

File: tests/baselines/reference/declarationEmitThisPredicatesWithPrivateName01.js [hunk 2]
```diff
@@ -37,3 +37,12 @@ var D = /** @class */ (function (_super) {
...
 }(C));
+
+
+//// [declarationEmitThisPredicatesWithPrivateName01.d.ts]
+export declare class C {
+    m(): this is D;
+}
+declare class D extends C {
+}
+export {};
```
[the remark this answers is not in the document]
Comment (weswigham) [reply tests/baselines/reference/declarationEmitThisPredicatesWithPrivateName01.js]:
```
declare class D extends C {
}
```
isn't exported (but it used in `C`). To keep it that way, we need to emit `export {}` to fix the scoping of the file.

[the remark this answers is not in the document]
Comment (sheetalkamat) [reply tests/baselines/reference/declarationEmitThisPredicatesWithPrivateName01.js]: But there is export declare class C so export {} is not needed. https://github.com/Microsoft/TypeScript/pull/23351/files/aa71540de296549fa21676bee32349088189fa35#diff-880f6135e59878559ff8c952515ec352R43

[the remark this answers is not in the document]
Comment (sheetalkamat) [reply tests/baselines/reference/declarationEmitThisPredicatesWithPrivateName01.js]: In mostly all of these scenarios the error you are fixing is public -> private so you are just adding private types to file, the scoping of file shouldnt affect by any of these additions.

[the remark this answers is not in the document]
Comment (weswigham) [reply tests/baselines/reference/declarationEmitThisPredicatesWithPrivateName01.js]: No, it is. Our scope rules are odd any `export` statement makes it a module, _but_ in a `.d.ts` unless there's an export assignment or export declaration, _every_ top-level declaration is seen at exported, even _without_ an explicit export modifier. AFAIK this was a decision made to reduce the noise in writing declaration files a long time ago.

**Comment by bluelovers** [comment]
does can fix this code too? and `@ts-ignore` not work here

```ts
function fn(inputRegExp: RegExp, input: string){}

namespace fn
{
	export const S = Symbol('fn');
}

// @ts-ignore
const fn2 = fn;

export = fn2;
```

> Error:(13, 7) TS4025: Exported variable 'fn2' has or is using private name 'fn'.

File: src/compiler/transformers/declarations.ts [hunk 3]
```diff
@@ -1171,7 +1156,7 @@ namespace ts {
         }
 
         function ensureAccessor(node: AccessorDeclaration): PropertyDeclaration | undefined {
-            const accessors = getAllAccessorDeclarations((node.parent as ClassDeclaration).members, node);
+            const accessors = resolver.getAllAccessorDeclarations(node);
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/transformers/declarations.ts]: Computed property names have differing names and need to merge - the only way to do so is via their symbol.

**Comment by weswigham** [comment]
@rbuckton anything else?

**Pull request merged**